### ISSUE TITLE
Obscure errors thrown

### ISSUE THREAD

**Issue by theofidry** [psalm/psalm-plugin-phpunit#105]
I had a provider with this signature:

```php
public function providePropertyValue(): ?Generator
```

which although is not recommended is not incorrect either. The result was:

```
Uncaught Exception: should be iterable
Stack trace in the forked worker:
/path/to/vendor/psalm/plugin-phpunit/src/Hooks/TestCaseHandler.php
```

I think there is two issues here:
- the above should not result in a crash
- I can see 3 RuntimeException thrown in `TestCaseHandler` which a simple message: "Unexpected: union has no way to get it constituent types", "should be iterable", "unexpected type": IMO it's not very user-friendly to throw those exceptions like without any more context... IMO requiring the user to have xdebug, disallow the xdebug handler (all of that is harder if using the PHAR too) and then debugging this just to be able to report the issue is making it a bit harder than necessary...

**Comment by weirdan** [comment]
>  is not incorrect either

Does PHPUnit allow returning null from a data provider?

**Comment by weirdan** [comment]
> the above should not result in a crash

Sure. However that was something unexpected - thus the exception.

> IMO it's not very user-friendly to throw those exceptions like without any more context... IMO requiring the user to have xdebug, disallow the xdebug handler (all of that is harder if using the PHAR too) and then debugging this just to be able to report the issue is making it a bit harder than necessary...

Exceptions are not user-friendly by definition. Generally you just copy/paste the exception trace into an issue and authors can take it from there.

As to why you didn't get the trace, though... it's interesting. Dumping exception trace for unhandled exception is the default PHP behavior, so something must be overriding that. It's not Psalm and not the plugin. Do you have autoloader specified in `psalm.xml`?

**Comment by theofidry** [comment]
>As to why you didn't get the trace, though... it's interesting. Dumping exception trace for unhandled exception is the default PHP behavior, so something must be overriding that. It's not Psalm and not the plugin. Do you have autoloader specified in psalm.xml?

I get a semblance of a trace I just putted the first line, the problem IMO is that you get an error and don't know what causes the failure which makes it harder to report

**Comment by theofidry** [comment]
>Does PHPUnit allow returning null from a data provider?

No idea to be honest... It was an incorrect signature but it was not returning null ever

**Comment by weirdan** [comment]
Please try with `dev-master` - I've just merged the change that should fix the issue.

**Comment by theofidry** [comment]
In the scenario above it would not fail anymore, but if you have a non-iterable signature it will still fail right?

>Exceptions are not user-friendly by definition. Generally you just copy/paste the exception trace into an issue and authors can take it from there.

Yes, but at the same time providing more information both helps:

- knowing where the issue is for a temporary workaround if necessary
- reporting the issue with a reproducer

For example if in the failure above I would have had something like `unexpected PHPUnit data provider return type for Acme\MyTest::providePropertyValue()` it would have cut the debug time by 3 and make it easier to provide a reproducer

**Comment by weirdan** [comment]
> In the scenario above it would not fail anymore, but if you have a non-iterable signature it will still fail right?

No. There's a check before the call to `unionizeIterables()` that reports non-iterable return type and skips to the next data provider. It's that skip that was incorrect (`continue` vs `continue 2`).

**Comment by theofidry** [comment]
Ha I see, LGTM then! Thanks for the quick fix :)

**Comment by weirdan** [comment]
Fix released in [0.15.1](https://github.com/psalm/psalm-plugin-phpunit/releases/tag/0.15.1).
